lib/compiler/src/engine/link.rs

```diff
@@ -130,18 +131,28 @@ fn apply_relocation(
 
 /// Links a module, patching the allocated functions with the
 /// required relocations and jump tables.
-pub fn link_module(
+pub fn link_module<'a>(
     _module: &ModuleInfo,
     allocated_functions: &PrimaryMap<LocalFunctionIndex, FunctionExtent>,
-    function_relocations: Relocations,
+    function_relocations: impl Iterator<
```

Comment:
nit: this is very noisy in the argument signature, can we move that to a generic bound?

Reply:
The generic version of this function definition is much less readable IMO:

```
pub fn link_module<
    'a,
    FunctionRelocationIterator,
    FunctionRelocationIteratorInner,
    FunctionRelocation,
    SectionRelocationIterator,
    SectionRelocationIteratorInner,
    SectionRelocation,
>(
    _module: &ModuleInfo,
    allocated_functions: &PrimaryMap<LocalFunctionIndex, FunctionExtent>,
    function_relocations: FunctionRelocationIterator,
    allocated_sections: &PrimaryMap<SectionIndex, SectionBodyPtr>,
    section_relocations: SectionRelocationIterator,
    libcall_trampolines: SectionIndex,
    trampoline_len: usize,
) where
    FunctionRelocationIterator:
        Iterator<Item = (LocalFunctionIndex, FunctionRelocationIteratorInner)>,
    FunctionRelocationIteratorInner: Iterator<Item = &'a FunctionRelocation>,
    FunctionRelocation: RelocationLike + 'a,
    SectionRelocationIterator: Iterator<Item = (SectionIndex, SectionRelocationIteratorInner)>,
    SectionRelocationIteratorInner: Iterator<Item = &'a SectionRelocation>,
    SectionRelocation: RelocationLike + 'a,
```

Reply:
I just meant the one large type.

But that of course also is messy because of one generic vs 5 `impls`...

No objection to leaving it as is.